File: backend/src/main/java/ch/rodano/configuration/model/scope/ScopeModel.java
package ch.rodano.configuration.model.scope;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.SortedMap;
import java.util.SortedSet;
import java.util.TreeMap;
import java.util.TreeSet;
import java.util.stream.Collectors;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonInclude.Include;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import com.fasterxml.jackson.annotation.JsonPropertyOrder;

import ch.rodano.configuration.exceptions.NoNodeException;
import ch.rodano.configuration.exceptions.NoRespectForConfigurationException;
import ch.rodano.configuration.model.cms.CMSLayout;
import ch.rodano.configuration.model.common.Entity;
import ch.rodano.configuration.model.common.Node;
import ch.rodano.configuration.model.common.SuperDisplayable;
import ch.rodano.configuration.model.dataset.DatasetModel;
import ch.rodano.configuration.model.event.EventGroup;
import ch.rodano.configuration.model.event.EventModel;
import ch.rodano.configuration.model.form.FormModel;
import ch.rodano.configuration.model.payment.PayableModel;
import ch.rodano.configuration.model.profile.Profile;
import ch.rodano.configuration.model.reports.WorkflowStatesSelector;
import ch.rodano.configuration.model.rights.RightAssignable;
import ch.rodano.configuration.model.rules.Rule;
import ch.rodano.configuration.model.study.Study;
import ch.rodano.configuration.model.timelinegraph.TimelineGraph;
import ch.rodano.configuration.model.workflow.Workflow;
import ch.rodano.configuration.model.workflow.WorkflowState;
import ch.rodano.configuration.model.workflow.WorkflowableModel;
import ch.rodano.configuration.utils.DisplayableUtils;

@JsonInclude(Include.NON_NULL)
@JsonPropertyOrder(alphabetic = true)
public class ScopeModel implements Serializable, SuperDisplayable, WorkflowableModel, PayableModel, Node, RightAssignable<ScopeModel> {
	private static final long serialVersionUID = -3652751984945692998L;

	public static final Comparator<ScopeModel> DEFAULT_COMPARATOR = Comparator.comparing(ScopeModel::getId);

	public static final Comparator<ScopeModel> COMPARATOR_DEPTH = (s1, s2) -> {
		final var s1Depth = s1.getDepth();
		final var s2Depth = s2.getDepth();
		if(s1Depth == s2Depth) {
			return s1.getId().compareTo(s2.getId());
		}
		return s1Depth - s2Depth;
	};

	private Study study;
	private String id;

	private SortedMap<String, String> shortname;
	private SortedMap<String, String> longname;
	private SortedMap<String, String> description;
	private SortedMap<String, String> pluralShortname;

	private String defaultParentId;
	private List<String> parentIds;
	private boolean virtual;

	private Integer maxNumber;
	private Integer expectedNumber;
	private String scopeFormat;

	private String defaultProfileId;

	private List<EventGroup> eventGroups;
	private List<EventModel> eventModels;

	private List<String> datasetModelIds;
	private List<String> formModelIds;
	private List<String> workflowIds;

	private List<WorkflowStatesSelector> workflowStatesSelectors;

	private CMSLayout layout;

	private List<Rule> createRules;
	private List<Rule> removeRules;
	private List<Rule> restoreRules;

	public ScopeModel() {
		shortname = new TreeMap<>();
		longname = new TreeMap<>();
		description = new TreeMap<>();
		pluralShortname = new TreeMap<>();
		parentIds = new ArrayList<>();
		eventGroups = new ArrayList<>();
		eventModels = new ArrayList<>();
		workflowIds = new ArrayList<>();
		datasetModelIds = new ArrayList<>();
		formModelIds = new ArrayList<>();
	}

	@JsonBackReference
	public final void setStudy(final Study study) {
		this.study = study;
	}

	@JsonBackReference
	public final Study getStudy() {
		return study;
	}

	@Override
	public final String getId() {
		return id;
	}

	public final void setId(final String id) {
		this.id = id;
	}

	@Override
	public final SortedMap<String, String> getDescription() {
		return description;
	}

	public void setDescription(final SortedMap<String, String> description) {
		this.description = description;
	}

	@Override
	public final SortedMap<String, String> getLongname() {
		return longname;
	}

	public final void setLongname(final SortedMap<String, String> longname) {
		this.longname = longname;
	}

	@Override
	public final SortedMap<String, String> getShortname() {
		return shortname;
	}

	public final void setShortname(final SortedMap<String, String> shortname) {
		this.shortname = shortname;
	}

	public final SortedMap<String, String> getPluralShortname() {
		return pluralShortname;
	}

	public final void setPluralShortname(final SortedMap<String, String> pluralShortname) {
		this.pluralShortname = pluralShortname;
	}

	@JsonIgnore
	public final String getLocalizedPluralShortname(final String... languages) {
		final var plural = DisplayableUtils.getLocalizedMap(pluralShortname, languages);
		if(plural == null || plural.isEmpty()) {
			return getLocalizedShortname(languages);
		}
		return plural;
	}

	@JsonIgnore
	public String getIdAndLocalizedShortname(final String... languages) {
		final var localizedShortname = getLocalizedShortname(languages);
		if(id.equals(localizedShortname)) {
			return localizedShortname;
		}
		return String.format("%s (%s)", id, localizedShortname);
	}

	public final String getDefaultParentId() {
		return defaultParentId;
	}

	public final void setDefaultParentId(final String defaultParentId) {
		this.defaultParentId = defaultParentId;
	}

	public final List<String> getParentIds() {
		return parentIds;
	}

	public final void setParentIds(final List<String> parentIds) {
		this.parentIds = parentIds;
	}

	public final void setVirtual(final boolean virtual) {
		this.virtual = virtual;
	}

	public final boolean isVirtual() {
		return virtual;
	}

	public final Integer getMaxNumber() {
		return maxNumber;
	}

	public final void setMaxNumber(final Integer maxNumber) {
		this.maxNumber = maxNumber;
	}

	public final Integer getExpectedNumber() {
		return expectedNumber;
	}

	public final void setExpectedNumber(final Integer exptectedNumber) {
		expectedNumber = exptectedNumber;
	}

	@JsonManagedReference
	public void setEventModels(final List<EventModel> eventModels) {
		this.eventModels = eventModels;
	}

	@JsonManagedReference
	public List<EventModel> getEventModels() {
		return eventModels;
	}

	@JsonIgnore
	public EventModel getEventModel(final String eventModelId) {
		return eventModels.stream()
			.filter(e -> e.getId().equalsIgnoreCase(eventModelId))
			.findAny()
			.orElseThrow(() -> new NoNodeException(this, Entity.EVENT_MODEL, eventModelId));
	}

	@JsonManagedReference
	public final void setEventGroups(final List<EventGroup> eventGroups) {
		this.eventGroups = eventGroups;
	}

	@JsonManagedReference
	public final List<EventGroup> getEventGroups() {
		return eventGroups;
	}

	@JsonIgnore
	public EventGroup getEventGroup(final String eventGroupId) {
		return eventGroups.stream()
			.filter(e -> e.getId().equalsIgnoreCase(eventGroupId))
			.findAny()
			.orElseThrow(() -> new NoNodeException(this, Entity.EVENT_GROUP, eventGroupId));
	}

	public final List<String> getDatasetModelIds() {
		return datasetModelIds;
	}

	public final void setDatasetModelIds(final List<String> datasetModelIds) {
		this.datasetModelIds = datasetModelIds;
	}

	public final List<String> getFormModelIds() {
		return formModelIds;
	}

	public final void setFormModelIds(final List<String> formModelIds) {
		this.formModelIds = formModelIds;
	}

	public final String getDefaultProfileId() {
		return defaultProfileId;
	}

	public final void setDefaultProfileId(final String defaultProfileId) {
		this.defaultProfileId = defaultProfileId;
	}

	public List<WorkflowStatesSelector> getWorkflowStatesSelectors() {
		return workflowStatesSelectors;
	}

	public void setWorkflowStatesSelectors(final List<WorkflowStatesSelector> workflowStatesSelectors) {
		this.workflowStatesSelectors = workflowStatesSelectors;
	}

	public final CMSLayout getLayout() {
		return layout;
	}

	public final void setLayout(final CMSLayout layout) {
		this.layout = layout;
	}

	public final String getScopeFormat() {
		return scopeFormat;
	}

	public final void setScopeFormat(final String scopeFormat) {
		this.scopeFormat = scopeFormat;
	}

	public List<Rule> getCreateRules() {
		return createRules;
	}

	public void setCreateRules(final List<Rule> createRules) {
		this.createRules = createRules;
	}

	public List<Rule> getRemoveRules() {
		return removeRules;
	}

	public void setRemoveRules(final List<Rule> removeRules) {
		this.removeRules = removeRules;
	}

	public List<Rule> getRestoreRules() {
		return restoreRules;
	}

	public void setRestoreRules(final List<Rule> restoreRules) {
		this.restoreRules = restoreRules;
	}

	@JsonIgnore
	public final boolean hasLayout() {
		return layout != null;
	}

	@JsonIgnore
	public ScopeModel getDefaultParent() {
		if(isRoot()) {
			throw new NoNodeException(Entity.SCOPE_MODEL, String.format("No default parent for root scope model %s", getId()));
		}
		return getStudy().getScopeModel(defaultParentId);
	}

	@JsonIgnore
	public List<ScopeModel> getDefaultAncestors() {
		if(isRoot()) {
			return new ArrayList<>();
		}
		final List<ScopeModel> ancestors = new ArrayList<>();
		final var defaultParent = getDefaultParent();
		ancestors.add(defaultParent);
		ancestors.addAll(defaultParent.getDefaultAncestors());
		return ancestors;
	}

	@JsonIgnore
	public List<ScopeModel> getScopeModelParents() {
		return getParentIds().stream().map(study::getScopeModel).toList();
	}

	@JsonIgnore
	public List<ScopeModel> getScopeModelAncestors() {
		final List<ScopeModel> ancestors = new ArrayList<>();
		getScopeModelParents().forEach(parent -> {
			ancestors.add(parent);
			ancestors.addAll(parent.getScopeModelAncestors());
		});
		return ancestors;
	}

	@JsonIgnore
	public List<ScopeModel> getChildrenScopeModel() {
		return getStudy().getScopeModels().stream()
			.filter(s -> s.getParentIds().contains(getId()))
			.toList();
	}

	@JsonIgnore
	public final List<ScopeModel> getDescendantsScopeModel() {
		final List<ScopeModel> descendants = new ArrayList<>();
		for(final var child : getChildrenScopeModel()) {
			descendants.add(child);
			descendants.addAll(child.getDescendantsScopeModel());
		}
		return descendants;
	}

	@JsonIgnore
	public final List<ScopeModel> getBranch(final ScopeModel model) {
		final List<ScopeModel> models = new ArrayList<>();
		models.add(this);

		if(getParentIds().isEmpty()) {
			throw new NoRespectForConfigurationException(String.format("%s and %s are not on the same branch", models.get(0).getId(), model.getId()));
		}

		if(getParentIds().contains(model.getId())) {
			models.add(model);
		}
		else {
			for(final var parent : getScopeModelParents()) {
				for(final var branch : parent.getBranch(model)) {
					if(!models.contains(branch)) {
						models.add(branch);
					}
				}
			}
		}
		return models;
	}

	@Override
	@JsonIgnore
	public final int compareTo(final ScopeModel otherScopeModel) {
		return DEFAULT_COMPARATOR.compare(this, otherScopeModel);
	}

	@JsonIgnore
	public final boolean isRoot() {
		return parentIds.isEmpty();
	}

	@JsonIgnore
	public final boolean isLeaf() {
		return getStudy().getScopeModels().stream().noneMatch(s -> s.parentIds.contains(id));
	}

	@JsonIgnore
	public final int getDepth() {
		if(isRoot()) {
			return 0;
		}
		return 1 + getDefaultParent().getDepth();
	}

	@JsonIgnore
	public final boolean isAncestorOf(final ScopeModel potentialDescendant) {
		return potentialDescendant.isDescendantOf(this);
	}

	@JsonIgnore
	public final boolean isDescendantOf(final ScopeModel potentialAncestor) {
		if(isChildOf(potentialAncestor)) {
			return true;
		}
		for(final var s : getScopeModelParents()) {
			if(s.isDescendantOf(potentialAncestor)) {
				return true;
			}
		}
		return false;
	}

	@JsonIgnore
	public final boolean isParentOf(final ScopeModel potentialChild) {
		return potentialChild.getParentIds().contains(getId());
	}

	@JsonIgnore
	public final boolean isChildOf(final ScopeModel potentialParent) {
		return getParentIds().contains(potentialParent.getId());
	}

	@JsonIgnore
	public final boolean isDescendantOf(final List<ScopeModel> potentialAncestors) {
		return potentialAncestors.stream()
			.anyMatch(this::isDescendantOf);
	}

	@JsonIgnore
	public List<DatasetModel> getDatasetModels() {
		return datasetModelIds.stream().map(study::getDatasetModel).toList();
	}

	@JsonIgnore
	public List<FormModel> getFormModels() {
		return study.getNodesFromIds(Entity.FORM_MODEL, formModelIds);
	}

	@Override
	public final List<String> getWorkflowIds() {
		return workflowIds;
	}

	public final void setWorkflowIds(final List<String> workflowIds) {
		this.workflowIds = workflowIds;
	}

	@Override
	@JsonIgnore
	public final List<Workflow> getWorkflows() {
		return getStudy().getNodesFromIds(Entity.WORKFLOW, getWorkflowIds());
	}

	@Override
	public final Entity getEntity() {
		return Entity.SCOPE_MODEL;
	}

	@Override
	@JsonIgnore
	public final Collection<Node> getChildrenWithEntity(final Entity entity) {
		switch(entity) {
			case SCOPE_MODEL:
				return Collections.unmodifiableList(getChildrenScopeModel());
			case EVENT_GROUP:
				return Collections.unmodifiableList(getEventGroups());
			case EVENT_MODEL:
				return Collections.unmodifiableList(getEventModels());
			default:
				return Collections.emptyList();
		}
	}

	@JsonIgnore
	public final List<EventModel> getInceptiveEventModels() {
		return getEventModels().stream()
			.filter(EventModel::isInceptive)
			.toList();
	}

	@JsonIgnore
	public Optional<Profile> getDefaultProfile() {
		return Optional.ofNullable(defaultProfileId).map(study::getProfile);
	}

	@JsonIgnore
	public void setDefaultProfile(final Profile profile) {
		setDefaultProfileId(profile == null ? null : profile.getId());
	}

	@JsonIgnore
	public SortedSet<TimelineGraph> getGraphConfigs() {
		return study.getTimelineGraphs().stream()
			.filter(conf -> conf.getScopeModelId().equals(getId()))
			.collect(Collectors.toCollection(TreeSet::new));
	}

	@JsonIgnore
	public List<WorkflowState> getWorkflowStates() {
		final List<WorkflowState> states = new ArrayList<>();
		for(final WorkflowStatesSelector selector : workflowStatesSelectors) {
			final var workflow = study.getWorkflow(selector.getWorkflowId());
			selector.getStateIds().stream().map(workflow::getState).forEach(states::add);
		}
		return states;
	}

	@Override
	@JsonIgnore
	public String getAssignableDescription() {
		return getDefaultLocalizedShortname();
	}

	@JsonIgnore
	public final String getDefaultLocalizedShortname() {
		return getLocalizedShortname(study.getDefaultLanguage().getId());
	}
}

File: frontends/configurator/www/model/config/entities/scope_model.js
import '../../../basic-tools/extension.js';

import {DisplayableNode} from '../node_displayable.js';
import {EntitiesHooks} from '../entities_hooks.js';
import {Study} from './study.js';
import {Entities} from '../entities.js';
import {RuleEntities} from '../rule_entities.js';
import {Utils} from '../utils.js';

export class ScopeModel extends DisplayableNode {
	static getProperties() {
		return {
			study: {type: Entities.Study.name, back_reference: true},
			id: {type: 'string'},
			shortname: {type: 'object'},
			pluralShortname: {type: 'object'},
			longname: {type: 'object'},
			description: {type: 'object'},
			defaultParentId: {type: 'string'},
			parentIds: {type: 'array'},
			virtual: {type: 'boolean'},
			maxNumber: {type: 'number'},
			expectedNumber: {type: 'string'},
			scopeFormat: {type: 'string'},
			eventGroups: {type: 'array', subtype: Entities.EventGroup.name},
			eventModels: {type: 'array', subtype: Entities.EventModel.name},
			datasetModelIds: {type: 'array'},
			formModelIds: {type: 'array'},
			workflowIds: {type: 'array'},
			defaultProfileId: {type: 'string'},
			layout: {type: 'string'},
			workflowStatesSelectors: {type: 'array'},
			createRules: {type: 'array', subtype: Entities.Rule.name},
			removeRules: {type: 'array', subtype: Entities.Rule.name},
			restoreRules: {type: 'array', subtype: Entities.Rule.name}
		};
	}
	static RuleEntities = [
		RuleEntities.SCOPE
	];

	constructor(values) {
		super();
		this.study = undefined;
		this.id = undefined;
		this.shortname = {};
		this.pluralShortname = {};
		this.longname = {};
		this.description = {};
		this.defaultParentId = undefined;
		this.parentIds = [];
		this.virtual = false;
		this.maxNumber = undefined;
		this.expectedNumber = undefined;
		this.scopeFormat = undefined;
		this.eventGroups = [];
		this.eventModels = [];
		this.datasetModelIds = [];
		this.formModelIds = [];
		this.workflowIds = [];
		this.defaultProfileId = undefined;
		this.layout = undefined;
		this.workflowStatesSelectors = [];
		this.createRules = [];
		this.removeRules = [];
		this.restoreRules = [];
		EntitiesHooks?.CreateNode.call(this, values);
	}

	getScopeModelDefaultParent() {
		return this.defaultParentId ? this.study.getScopeModel(this.defaultParentId) : undefined;
	}
	getScopeModelParents() {
		return this.parentIds.map(Study.prototype.getScopeModel, this.study);
	}
	getScopeModelAncestors() {
		const ancestors = [];
		const parents = this.getScopeModelParents();
		ancestors.pushAll(parents);
		parents.forEach(function(parent) {
			parent.getScopeModelAncestors().forEach(function(ancestor) {
				if(!ancestors.includes(ancestor)) {
					ancestors.push(ancestor);
				}
			});
		});
		return ancestors;
	}
	getScopeModelChildren() {
		const scope_model_id = this.id;
		return this.study.scopeModels.filter(s => s.parentIds.includes(scope_model_id));
	}
	getScopeModelDescendants() {
		const scope_model = this;
		return this.study.scopeModels.filter(s => s.getScopeModelAncestors().includes(scope_model));
	}
	getScopeModelBranch(scope_model) {
		const scope_models = [];
		scope_models.push(this);
		if(this !== scope_model) {
			if(this.parentIds.isEmpty()) {
				throw new Error(`${scope_models[0].id} and ${scope_model.id} are not on the same branch`);
			}
			if(this.parentIds.includes(scope_model.id)) {
				scope_models.push(scope_model);
			}
			else {
				const parent_scope_models = this.getScopeModelParents();
				for(let i = 0; i < parent_scope_models.length; i++) {
					const branch_scope_models = parent_scope_models[i].getScopeModelBranch(scope_model);
					for(let j = 0; j < branch_scope_models.length; j++) {
						const branch_scope_model = branch_scope_models[j];
						if(!scope_models.includes(branch_scope_model)) {
							scope_models.push(branch_scope_model);
						}
					}
				}
			}
		}
		return scope_models;
	}
	isRoot() {
		return this.parentIds.isEmpty();
	}
	isLeaf() {
		return this.study.getLeafScopeModel().id === this.id;
	}
	getEventModel(event_model_id) {
		return Utils.getObjectById(this.eventModels, event_model_id);
	}
	getEventGroup(event_group_id) {
		return Utils.getObjectById(this.eventGroups, event_group_id);
	}
	getWorkflows() {
		return this.workflowIds.map(Study.prototype.getWorkflow, this.study);
	}
	getDatasetModels() {
		return this.datasetModelIds.map(Study.prototype.getDatasetModel, this.study);
	}
	getFormModels() {
		return this.formModelIds.map(Study.prototype.getFormModel, this.study);
	}
	getInceptiveEventModel() {
		const inceptive_event_model = this.eventModels.find(e => e.inceptive);
		if(inceptive_event_model) {
			return inceptive_event_model;
		}
		else {
			throw new Error(`No inceptive event model for scope model ${this.id}`);
		}
	}

	//rulable and layoutable
	getStudy() {
		return this.study;
	}

	//tree
	getChildren(entity, index) {
		switch(entity) {
			case Entities.EventGroup:
				return this.eventGroups.slice();
			case Entities.EventModel:
				return this.eventModels.slice();
			case Entities.CMSLayout:
				return this.layout ? [this.layout] : [];
			case Entities.Rule:
				if(index === 0) {
					return this.createRules.slice();
				}
				else if(index === 1) {
					return this.removeRules.slice();
				}
				else if(index === 2) {
					return this.restoreRules.slice();
				}
		}
		throw new Error(`Entity ${entity.name} is not a child of entity ${this.getEntity().name}`);
	}
	addChild(child, index) {
		const child_entity = child.getEntity();
		switch(child_entity) {
			case Entities.EventGroup:
				this.eventGroups.push(child);
				child.scopeModel = this;
				break;
			case Entities.EventModel:
				this.eventModels.push(child);
				child.scopeModel = this;
				break;
			case Entities.Rule:
				if(index === 0) {
					this.createRules.push(child);
				}
				else if(index === 1) {
					this.removeRules.push(child);
				}
				else if(index === 2) {
					this.restoreRules.push(child);
				}
				child.rulable = this;
				break;
			default:
				throw new Error(`Entity ${child_entity.name} cannot be added as a child of entity ${this.getEntity().name}`);
		}
		EntitiesHooks?.AddChildNode.call(this, child);
	}
	getRelations(entity) {
		switch(entity) {
			case Entities.DatasetModel:
				return this.getDatasetModels();
			case Entities.FormModel:
				return this.getFormModels();
			case Entities.Workflow:
				return this.getWorkflows();
		}
		throw new Error(`Entity ${entity.name} is not related to entity ${this.getEntity().name}`);
	}
	//entity scope model is an exception because it is not possible to know if it is used based on the configuration
	//TODO in theory, "isUsed" should be consistent with "getUsage", this is not the case with scope models
	isUsed () {
		return true;
	}

	//bus
	onChangeScopeModelId(event) {
		this.parentIds.replace(event.oldValue, event.newValue);
		if(this.defaultParentId && this.defaultParentId === event.oldValue) {
			this.defaultParentId = event.newValue;
		}
	}
	onDeleteScopeModel(event) {
		this.parentIds.removeElement(event.node.id);
		if(this.defaultParentId === event.node.id) {
			this.defaultParentId = undefined;
		}
	}

	onDeleteEventGroup(event) {
		this.eventGroups.removeElement(event.node);
	}
	onMoveEventGroup(event) {
		if(event.newParent === this) {
			event.node.scopeModel.eventGroups.removeElement(event.node);
			event.node.scopeModel = this;
			this.eventGroups.push(event.node);
		}
	}

	onDeleteEventModel(event) {
		this.eventModels.removeElement(event.node);
	}
	onMoveEventModel(event) {
		if(event.newParent === this) {
			event.node.scopeModel.eventModels.removeElement(event.node);
			event.node.scopeModel = this;
			this.eventModels.push(event.node);
		}
	}

	onChangeDatasetModelId(event) {
		this.datasetModelIds.replace(event.oldValue, event.newValue);
	}
	onDeleteDatasetModel(event) {
		this.datasetModelIds.removeElement(event.node.id);
	}

	onChangeFormModelId(event) {
		this.formModelIds.replace(event.oldValue, event.newValue);
	}
	onDeleteFormModel(event) {
		this.formModelIds.removeElement(event.node.id);
	}

	onChangeWorkflowId(event) {
		this.workflowIds.replace(event.oldValue, event.newValue);
	}
	onDeleteWorkflow(event) {
		this.workflowIds.removeElement(event.node.id);
	}

	onDeleteWorkflowStatesSelector(event) {
		this.workflowStatesSelectors.removeElement(event.node);
	}
	onMoveWorkflowStatesSelector(event) {
		if(event.newParent === this) {
			event.node.parent.workflowStatesSelectors.removeElement(event.node);
			event.node.parent = this;
			this.workflowStatesSelectors.push(event.node);
		}
	}

	onChangeProfileId(event) {
		if(this.defaultProfileId && this.defaultProfileId === event.oldValue) {
			this.defaultProfileId = event.newValue;
		}
	}
	onDeleteProfile(event) {
		if(this.defaultProfileId === event.node.id) {
			this.defaultProfileId = undefined;
		}
	}

	onDeleteRule(event) {
		this.createRules.removeElement(event.node);
		this.removeRules.removeElement(event.node);
		this.restoreRules.removeElement(event.node);
	}
}
